### ISSUE TITLE
Add `JsonSchema` Support for Multiple Types

### ISSUE THREAD

**Issue by solonovamax** [kotest/kotest#4463]
Basically just JSON Schema's [`anyOf`](https://json-schema.org/understanding-json-schema/reference/combining#anyOf) and/or [`oneOf`](https://json-schema.org/understanding-json-schema/reference/combining#oneOf)

This feature is similar, although not identical, to #2980. (in certain cases #2980 is a replacement for this, although it does not fit all cases)

an example syntax would look like this:
```kotlin
val stringOrIntegerListSchema = jsonSchema {
    array {
        anyOf(string(), integer())
    }
}

"""[ "foo", "bar", "baz" ]""" shouldMatchSchema stringOrIntegerListSchema // passes
"""[ 1, 2, 3 ]""" shouldMatchSchema stringOrIntegerListSchema // passes
"""[ true, false ]""" shouldMatchSchema stringOrIntegerListSchema // fails
```

in order to do this, it would either be necessary to add an additional `JsonSchemaElement` called something like `JsonCompositeElement`, which composes multiple elements (imo the less desirable choice)
or to change the `JsonSchema.JsonObjectBuilder.properties` field from a `MutableMap<String, JsonSchemaElement>` to a `MutableMap<String, Matcher<JsonSchemaElement>>` (imo the better choice)

**Comment by AlexCue987** [comment]
can you share some real life use cases where this feature would be needed?

**Comment by solonovamax** [comment]
> can you share some real life use cases where this feature would be needed?

I'm trying to make a schema that matches a [`fabric.mod.json`](https://fabricmc.net/wiki/documentation:fabric_mod_json_spec) file.
A couple of examples from it:
- the `mixins` field which is an array of elements that are either `String`s or an object with a `config` key and an optional `environment` key, the string being equivalent to just specifying the `config`.
- the `license` field which is either an array `String`s or a `String`. the string being equivalent to specifying a single element array.
- the `icon` field which is either a `String` or an object of `String` -> `String`. for the first case, the icon is used at all resolutions, in the second case the resolution of the icon is specified in the key.
- the `authors`/`contributors` field which is an array of either `String`s or an object with a `name` key and an optional `contact` key, the string being equivalent to just specifying the `name`
- the `entrypoints` field which is an object of `String` -> array of `String` or an object with a `value` and an optional `adapter` key, the string being equivalent to just specifying the `value`.

I don't have any control over this schema, and am just generating some json according to this schema and want to validate it.

**Comment by solonovamax** [comment]
My serialization code can be found here if that would help: [`SerialFabricModJson.kt`](https://github.com/solo-studios/nyx/blob/master/src/main/kotlin/ca/solostudios/nyx/internal/SerialFabricModJson.kt). This should be able to parse all valid `fabric.mod.json` files and fail on any invalid one.

**Comment by solonovamax** [comment]
Also, I sent a message in the kotlinlang's kotest slack channel, if you'd like to discuss this further there in a platform more suitable for Q&A

**Comment by AlexCue987** [comment]
Hey @LeoColman - would it make sense to add this functionality?

**Comment by solonovamax** [comment]
> Hey @LeoColman - would it make sense to add this functionality? 

We could discuss it in the kotlinlang slack, as it's more appropriate for serial communication (eg. conversations)

https://kotlinlang.slack.com/archives/CT0G9SD7Z/p1729824637519569

**Comment by sksamuel** [comment]
@AlexCue987 are you able to work on this and #2980 ?

**Comment by AlexCue987** [comment]
@sksamuel presumably this suggestion makes sense. Hi @solonovamax, because you already have a working solution, can you submit a PR?

**Comment by solonovamax** [comment]
> @sksamuel presumably this suggestion makes sense. Hi @solonovamax, because you already have a working solution, can you submit a PR?

I don't have a working solution for this, however if you'd like, I could investigate one. but, I'd most likely need to make some api changes to `JsonSchema` & friends to do so. however, I won't be able to work on this for the next week.

**Comment by sksamuel** [comment]
@Kantis you wrote a lot of this originally, any thoughts on this, and #2980

**Comment by sksamuel** [comment]
@PreAgile did you want to make a PR for this ?

**Comment by PreAgile** [comment]
> [@PreAgile](https://github.com/PreAgile) did you want to make a PR for this ?

  Yes, I’d like to work on this.

  I’m thinking of adding `JsonAnyOf` / `JsonOneOf` as new `JsonSchemaElement` subtypes with matching `anyOf()` /
  `oneOf()` builder functions, since that seems like the least invasive approach. I’ll dig into the validation/
  parsing flow and put together a PR.

**Comment by sksamuel** [comment]
Added you to the kotest org, if you're interested, just means you can make branches directly on kotest instead of your fork.

**Comment by PreAgile** [comment]
Yes, I'd like to work on this. I've opened #5807 with an implementation that adds `JsonAnyOf`/`JsonOneOf` as nested subtypes under `JsonSchema` — same pattern as `JsonEnum` in #5771.

**Approach:**
- `anyOf()` / `oneOf()` builder functions for the DSL
- Validation: `anyOf` passes if any candidate matches, `oneOf` passes if exactly one
- Parser: `anyOf`/`oneOf` keys are checked before `type`-based dispatch
- Includes tests for overlap semantics (e.g. `oneOf(number(), integer())` with `1` correctly fails)

Happy to adjust if you'd prefer a different direction.
